src/odemis/util/comp.py

```diff
@@ -141,10 +141,15 @@ def generate_zlevels(focuser: "Actuator", zrange: Tuple[float, float], zstep: fl
         raise KeyError(f"The focus actuator {focuser} does not have z axis")
 
     focuser_pos = focuser.position.value["z"]
+    focuser_rng = focuser.axes["z"].range
+    if (focuser_pos > focuser_rng[1]) or (focuser_pos < focuser_rng[0]):
+        raise ValueError(f"The current focus position {focuser_pos} is outside focus range {focuser_rng}. "
```

Comment:
Actually, from what I understand, lines 162-163 already *try* to handle this case. It's actually done in a more soft way, as it still allows to acquire one image (instead of forbidding acquisition here).
However, there, it return a float, instead of a list with 1 float. 
=> Drop your extra changes, and adjust line 163.